### ISSUE TITLE
RNNs do not have gradients while using DataParallel in 1.4.0

### ISSUE THREAD

**Issue by shaanchandra** [pytorch/pytorch#33552]
## 🐛 Bug

I have tried running RNNs (LSTMs and GRUs) on multiple GPUs using `nn.DataParallel()` and when I print/plot the gradients, they are 0. They are not `Nan` or `None` but `0.0`. I have run the same code with CNNs and it worked fine. The RNN is part of `model.parameters()` so that is also not an issue.
This issue exists only in 1.4.0. When I rolled back to 1.3.0 it worked fine.

## To Reproduce

Following is a self contained code. You can pass `multi=True` and `multi=False` to test the gradients on multiple or single GPU.
```
import torch

from torch import nn
from torch.nn.utils.rnn import pad_packed_sequence, pack_padded_sequence

import warnings
warnings.filterwarnings("ignore")
import torch.nn.functional as F

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

class BiLSTM(nn.Module):
    def __init__(self,):
        super(BiLSTM, self).__init__()
        self.lstm = nn.LSTM(10, 16, bidirectional = True, batch_first= True)
        self.classifier = nn.Linear(32, 1)

    def forward(self, embed, lengths):
        # sorted_len, sorted_idxs = torch.sort(lengths, descending =True)
        # embed = embed[ sorted_idxs ,: , :].to(device)
        sorted_len = lengths
        packed_embed = pack_padded_sequence(embed, sorted_len, batch_first = True, enforce_sorted=False).to(device)
        all_states, hidden_states = self.lstm(packed_embed)
        all_states, _ = pad_packed_sequence(all_states, batch_first = True)
        
        out = torch.cat((hidden_states[0][0], hidden_states[0][1]), dim = 1)
        out = self.classifier(out)
        return out.to(device)
    
def train(multi = False):
    seq_len = 5
    batch_size = 8
    input_dim = 10
    inp = torch.randn(batch_size, seq_len, input_dim).to(device)
    labels = torch.tensor([0,1,0,1,0,1,0,1])
    lens = torch.tensor([5,2,4,2,4,5,3,3]).to(device)
    
    model = BiLSTM().to(device)
    if multi:
        model = nn.DataParallel(model)
		
    optimizer = torch.optim.AdamW(model.parameters(), lr = 1e-4, weight_decay = 1e-3)
    criterion = nn.BCELoss()
    
    
    epochs = 1
    for i in range(epochs):
        preds = model(inp, lens)
        preds = F.sigmoid(preds)
        loss = criterion(preds,  labels.float().to(device))
        
        optimizer.zero_grad()
        loss.backward()
        torch.nn.utils.clip_grad_norm_(model.parameters(), 10)
        optimizer.step()
        
        
        print("Gradients info..\n")
        model_log = model.module if multi else model
        for name, param in model_log.named_parameters():
            print("param {} grad = {}".format(name, param.grad.data.view(-1)))
        
    
if __name__ == '__main__':
    train(multi = False)
```

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0): 1.4.0
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source): NA
 - Python version: 3.6.3
 - CUDA/cuDNN version: CUDA=10.0.13 / CuDNN = 7.4.2 
 - GPU models and configuration: Nvidia 1080ti
 - Any other relevant information:

## Additional context

NA

cc @ezyang @gchanan @zou3519

**Comment by ptrblck** [comment]
Thanks for the code snippet to reproduce this issue @shaanchandra !
The gradient also fails with this reduced code:
```python
device = 'cuda'
model = nn.LSTM(10, 16, bidirectional = False, batch_first=True)
model.to(device)

seq_len = 5
batch_size = 8
input_dim = 10
inp = torch.randn(batch_size, seq_len, input_dim).to(device)

# Switch
if True:
    model = nn.DataParallel(model)

output = model(inp)[0]
output.mean().backward()
print([p.grad for p in model.parameters()])
```

I'm currently debugging the `gather` operation to see, what might kill the gradients.

**Comment by ptrblck** [comment]
It seems that `gather` and `scatter` are properly forwarding the gradients.
Also, the input gets valid gradients, if created with `requires_grad=True`.

**Comment by shaanchandra** [comment]
Hi @ptrblck ,
Thanks for the update. 
And I agree. All the other components get the gradients (for example, in my snippet I included the classifier to show it has gradients but the RNN component does not).

I was working with a more complex architecture involving attention and RNNs at different levels, and all the RNNs at each level did not have gradients while all the other components at each level did.
